### PR TITLE
Sequence parallel perf updates

### PR DESCRIPTION
**Pull request by ericharper**
Changes made to apex to match megatron-lm gpt training perf.

Mainly using `_all_gather_base` and `_reduce_scatter_base`.

A few other minor fixes.

@erhoo82, @crcrpar, for viz.

### CODE REVIEW

File: apex/transformer/functional/fused_softmax.py [hunk 1]
```diff
@@ -162,7 +162,10 @@ def is_kernel_available(self, mask, b, np, sq, sk):
         if (
             self.scaled_masked_softmax_fusion  # user want to fuse
             and self.input_in_float16  # input must be fp16
-            and mask is not None  # mask tensor must not be None
+            and (
+                self.attn_mask_type == AttnMaskType.causal
+                or (self.attn_mask_type == AttnMaskType.padding and mask is not None)
+            )
```
Comment (crcrpar) [inline apex/transformer/functional/fused_softmax.py:168]: does this check happen to be what @yidong72 suggested?

**Review by crcrpar — APPROVED** [review]
changes seem to be
- a new check in softmax
- order of calculating sequence lengths for p2p in non interleaving pipeline model parallel
- `all_gather` and `reduce_scatter` to `_all_gather_base` and `_reduce_scatter_base`.

**Review by eqy — APPROVED** [review]
Tests look good on `dgx1v`, LGTM

File: apex/transformer/functional/fused_softmax.py (hunk 1 above)
Comment (ericharper) [reply apex/transformer/functional/fused_softmax.py:168]: @yidong72 has a PR open to NeMo for his changes: https://github.com/NVIDIA/NeMo/pull/4594

Comment (crcrpar) [reply apex/transformer/functional/fused_softmax.py:168]: then what is this check for? clarity?

**Pull request merged**